src/org/etools/j1939_84/controllers/part1/BroadcastValidator.java

```diff
@@ -0,0 +1,163 @@
+                listener.onResult("");
+                samplePackets.forEach(p -> listener.onResult(p.getPacket().toTimeString()));
+
+                addOutcome(listener,
+                           partNumber,
+                           stepNumber,
+                           Outcome.INFO,
+                           "Unable to determine period for PGN " + pgn + " from " + Lookup.getAddressName(
+                                   moduleSourceAddress));
+            } else {
+                Packet packet0 = samplePackets.get(0).getPacket();
+                Packet packet1 = samplePackets.get(1).getPacket();
+                Packet packet2 = samplePackets.get(2).getPacket();
+
+                listener.onResult("");
+                listener.onResult(packet0.toTimeString());
+                listener.onResult(packet1.toTimeString());
+                listener.onResult(packet2.toTimeString());
+
+                LocalDateTime t0 = packet0.getTimestamp();
+                LocalDateTime t1 = packet1.getTimestamp();
+                long diff1 = ChronoUnit.MILLIS.between(t0, t1);
+
+                LocalDateTime t2 = packet2.getTimestamp();
+                long diff2 = ChronoUnit.MILLIS.between(t1, t2);
+
+                PgnDefinition pgnDefinition = j1939DaRepository.findPgnDefinition(pgn);
+
+                long broadcastPeriod = pgnDefinition.getBroadcastPeriod();
+                double maxBroadcastPeriod = broadcastPeriod * 1.1;
+                double minBroadcastPeriod = broadcastPeriod * 0.9;
+
+                // b. Fail if any parameter is not broadcast within -10% of the fixed, specified broadcast period.
+                if (!pgnDefinition.isVariableBroadcast() && (diff1 < minBroadcastPeriod || diff2 < minBroadcastPeriod)) {
+                    addOutcome(listener,
+                               partNumber,
+                               stepNumber,
+                               Outcome.FAIL,
+                               "Broadcast of " + pgn + " by OBD Module " + Lookup.getAddressName(moduleSourceAddress) + " is less than 90% specified broadcast period.");
+                }
+
+                // b. Fail if any parameter is not broadcast within +10% of the fixed, specified broadcast period.
+                // c. Fail if any parameter in a variable period broadcast message exceeds 110% of its recommended broadcast period.
+                if (diff1 > maxBroadcastPeriod || diff2 > maxBroadcastPeriod) {
+                    addOutcome(listener,
+                               partNumber,
+                               stepNumber,
+                               Outcome.FAIL,
+                               "Broadcast of " + pgn + " by OBD Module " + Lookup.getAddressName(moduleSourceAddress) + " is beyond 110% specified broadcast period.");
+                }
+
+            }
+        }
+
+    }
+
+    public Map<Integer, List<GenericPacket>> buildPGNPacketsMap(List<GenericPacket> packets, int moduleAddress) {
```

Comment:
Does it make sense to keep the entire recording in memory like this, or instead should it be 
  bus.read(time,timeunits).collect(Collectors.toMap(p->p.getId(), new ArrayList<>(p),(a,b)->{a.addAll(b);return a;}));
so only the results are stored in memory?

Reply:
With Step 26, we need to listen to the bus for a period of time to see which packets are being broadcast.  Most typically, the wait time is going to be 3 seconds.  However, in the worse case, it could be 90 seconds. The maximum broadcast period that I could find is 30 seconds.  If there are several modules and we have to wait 3 seconds for each, it's more time consuming. 
So the trade off is listening for a period per module or listening once and saving them in memory.  I think we start here and see how it goes.  I do like sorting them by PGN up front.  If we need to make changes to this class, I'll keep that in mind as an improvement.